documentation/tutorial/01_maps_and_tiles.md

```diff
@@ -0,0 +1,157 @@
+
+  private
+
+  def default_graphic
+    return "ǂ"
+  end
+end
+```
+
+Here, we extend the `Goby::Tile` class, and replace the behavior of the
+`#default_graphic` method, so that it returns a different unicode symbol. We also
+make a change to the constructor so that `@graphic` is set to the unicode symbol
+returned from the invocation of `#default_graphic`. Now, when we instantiate an
+object of type `Grass`, and include it into our map, that tile will be represented by
+the `ǂ` character. This will provide a bit more visual context to the player of where
+they are on the map.
+
+So now let's say we want to customize this Grass tile a bit further by creating a
+grass tile that is not passable. We could create a new class called something like
+`BlockedGrass`, or we could further customize our existing `Grass` class. Let's go
+with the latter for now.
+
+```
+class Grass < Goby::Tile
+  def initialize(passable: true, seen: false, description: "", events: [], monsters: [])
+    @passable = passable
+    @seen = seen
+    @description = description
+    @events = events
+    @monsters = monsters
+    @graphic = set_graphic(passable)
+  end
+
+  private
+
+  def default_graphic
+    return "ǂ"
+  end
+
+  def set_graphic(passable)
+    passable ? default_graphic : "¥"
+  end
+end
+```
+
+So here what we've done is added another private method called `#set_graphic`.
+This method is responsible for setting the `@graphic` property of the `Grass`
+instance. When an instance is created that is passable, the "ǂ" character will
+represent that tile, but when an instance is created that is not passable, the "¥"
+character will be displayed instead.
+
+## Map Class
+
+So now we arrive at the point in which we want to assemble our tiles into something
+cohesive. This is the responsibility of the `Map` class. The `Map` class is quite
+simple, with just 3 properties and 3 methods. The properties are:
+
+- name (defaults to "Map")
+- tiles (defaults to [[Tile.new]])
+- music (defaults to nil)
+
+For now, we will ignore the methods as they aren't something that you will be
+actively working with at this point.
+
+Now that we know the properties of the `Map` class, we are ready to start
+constructing our own maps. Let's start by creating a map called `Meadow1`.
+
+```
+class Meadow1 < Map
+  def initialize
+    super(name: "Open Meadow")
+
+    # Define the main tiles on this map.
+    grass = Grass.new(description: "You are standing on some grass.")
+    blocked_grass = Grass.new(passable: false)
+
+    # Fill the map with "grass."
+    @tiles = Array.new(9)
+
+    normal_row = [grass, grass, grass, grass, grass]
+    blocked_row = [grass, blocked_grass, blocked_grass, blocked_grass, grass]
+
+    @tiles[0] = normal_row
+    @tiles[1] = normal_row
+    @tiles[2] = normal_row
+    @tiles[3] = blocked_row
+    @tiles[4] = blocked_row
+    @tiles[5] = blocked_row
+    @tiles[6] = normal_row
+    @tiles[7] = normal_row
+    @tiles[8] = normal_row
+
+  end
+end
```

Comment:
As it currently stands, each `Tile` will need to be "cloned" (see `Tile.clone` method) on the `Map`. Otherwise, each will refer to the same object and thus modifying one `grass` will modify every other `grass` (this becomes more important with `Event`s and `Monster`s). Once one moves on to understand the framework better, one might write it like this (comments optional):

```ruby
# 5 columns
5.times do |y|
  @tiles[y] = []
  # 9 rows
  9.times do |x|
    @tiles[y][x] = grass.clone
  end
end
```

Then for the blocked area, one could do something like this:

```ruby
# 3 columns
3.times do |y|
  # 3 rows
  3.times do |x|
    # Start in the top-left position at y-x coordinate (3,1).
    @tiles[y+3][x+1] = blocked_grass.clone
  end
end
```

This simplifies the single reference issue. However, I do agree that your way looks much more approachable to a beginner. You will need to write `grass.clone` and `blocked_grass.clone` in your `normal_row` and `blocked_row` for this to work. You may want to briefly explain why that's necessary.

When I say y-x coordinate, I just mean that conventionally `x` moves from left to right and `y` moves up and down. In 2D arrays, you can see that the `y` coordinate comes first, followed by the `x` coordinate.

Reply:
Thanks for this y-x clarification. I wondered why things were being described that way and not the more typical x-y way.